File: app/src/main/java/com/mirash/familiar/activity/edit/CredentialsEditActivity.java
package com.mirash.familiar.activity.edit;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.mirash.familiar.Const;
import com.mirash.familiar.R;
import com.mirash.familiar.db.Credentials;
import com.mirash.familiar.model.CredentialsModel;
import com.mirash.familiar.tool.EditResultAction;
import com.mirash.familiar.view.StyledTextInputLayout;

/**
 * @author Mirash
 */
public class CredentialsEditActivity extends AppCompatActivity implements Observer<Credentials> {
    private StyledTextInputLayout titleInputLayout;
    private StyledTextInputLayout linkInputLayout;
    private StyledTextInputLayout loginInputLayout;
    private StyledTextInputLayout emailInputLayout;
    private StyledTextInputLayout phoneInputLayout;
    private StyledTextInputLayout passwordInputLayout;
    private StyledTextInputLayout pinInputLayout;
    private StyledTextInputLayout detailsInputLayout;
    private CredentialsEditViewModel model;
    private MenuItem saveMenuItem;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_credentials_edit);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        titleInputLayout = findViewById(R.id.item_title_input);
        linkInputLayout = findViewById(R.id.item_link_input);
        loginInputLayout = findViewById(R.id.item_login_input);
        emailInputLayout = findViewById(R.id.item_email_input);
        phoneInputLayout = findViewById(R.id.item_phone_input);
        passwordInputLayout = findViewById(R.id.item_password_input);
        pinInputLayout = findViewById(R.id.item_pin_input);
        detailsInputLayout = findViewById(R.id.item_details_input);

        model = new ViewModelProvider(this).get(CredentialsEditViewModel.class);
        Bundle args = getIntent().getExtras();
        if (args != null) {
            if (args.containsKey(Const.KEY_ID)) {
                int id = getIntent().getIntExtra(Const.KEY_ID, 0);
                model.setCredentialsId(id);
                model.getCredentialsLiveData().observe(this, this);
            }
            int position = getIntent().getIntExtra(Const.KEY_POSITION, 0);
            model.setCredentialsPosition(position);
        }
        initSaveButtonStateObserver();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        model.removeCredentialsObserver(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (model.getCredentialsId() == null) {
            getMenuInflater().inflate(R.menu.credentials_save_menu, menu);
        } else {
            getMenuInflater().inflate(R.menu.credentials_edit_menu, menu);
        }
        saveMenuItem = menu.findItem(R.id.credentials_edit_save);
        model.getSaveButtonEnableStateLiveData().observe(this, enabled -> {
            saveMenuItem.setEnabled(enabled);
            saveMenuItem.getIcon().setAlpha(enabled ? 255 : 78);
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        } else if (id == R.id.credentials_edit_save) {
            saveCredentials();
        } else if (id == R.id.credentials_edit_delete) {
            deleteCredentials();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onChanged(Credentials credentials) {
        if (credentials == null) {
            setResult(RESULT_CANCELED);
            finish();
            return;
        }
        titleInputLayout.setTextWithNoAnimation(credentials.getTitle());
        linkInputLayout.setTextWithNoAnimation(credentials.getLink());
        loginInputLayout.setTextWithNoAnimation(credentials.getLogin());
        emailInputLayout.setTextWithNoAnimation(credentials.getEmail());
        phoneInputLayout.setTextWithNoAnimation(credentials.getPhone());
        passwordInputLayout.setTextWithNoAnimation(credentials.getPassword());
        pinInputLayout.setTextWithNoAnimation(credentials.getPin());
        detailsInputLayout.setTextWithNoAnimation(credentials.getDetails());
    }

    private void saveCredentials() {
        CredentialsModel credentialsModel = new CredentialsModel();
        credentialsModel.setTitle(titleInputLayout.getText());
        credentialsModel.setLink(linkInputLayout.getText());
        credentialsModel.setLogin(loginInputLayout.getText());
        credentialsModel.setEmail(emailInputLayout.getText());
        credentialsModel.setPhone(phoneInputLayout.getText());
        credentialsModel.setPassword(passwordInputLayout.getText());
        credentialsModel.setPin(pinInputLayout.getText());
        credentialsModel.setDetails(detailsInputLayout.getText());
        credentialsModel.setPosition(model.getCredentialsPosition());
        model.saveCredentials(credentialsModel);
        Intent data = new Intent();
        if (model.getCredentialsId() == null) {
            data.putExtra(Const.KEY_EDIT_RESULT_ACTION, EditResultAction.CREATE);
        } else {
            data.putExtra(Const.KEY_EDIT_RESULT_ACTION, EditResultAction.UPDATE);
        }
        data.putExtra(Const.KEY_POSITION, model.getCredentialsPosition());
        setResult(RESULT_OK, data);
        finish();
    }

    private void deleteCredentials() {
        model.deleteCredentials();
        Intent data = new Intent();
        data.putExtra(Const.KEY_POSITION, model.getCredentialsPosition());
        data.putExtra(Const.KEY_EDIT_RESULT_ACTION, EditResultAction.DELETE);
        setResult(RESULT_OK, data);
        finish();
    }

    private void initSaveButtonStateObserver() {
        initTextFillObserver(titleInputLayout, CredentialsEditViewModel.INDEX_TITLE);
        initTextFillObserver(linkInputLayout, CredentialsEditViewModel.INDEX_LINK);
        initTextFillObserver(loginInputLayout, CredentialsEditViewModel.INDEX_LOGIN);
        initTextFillObserver(emailInputLayout, CredentialsEditViewModel.INDEX_EMAIL);
        initTextFillObserver(phoneInputLayout, CredentialsEditViewModel.INDEX_PHONE);
        initTextFillObserver(passwordInputLayout, CredentialsEditViewModel.INDEX_PASSWORD);
        initTextFillObserver(pinInputLayout, CredentialsEditViewModel.INDEX_PIN);
    }

    private void initTextFillObserver(StyledTextInputLayout inputLayout, int index) {
        inputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(index, editable.length() > 0);
            }
        });
    }
}

File: app/src/main/java/com/mirash/familiar/activity/edit/CredentialsEditActivity.kt
package com.mirash.familiar.activity.edit

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.mirash.familiar.R
import com.mirash.familiar.databinding.ActivityCredentialsEditBinding
import com.mirash.familiar.db.Credentials
import com.mirash.familiar.model.credentials.CredentialsModel
import com.mirash.familiar.tool.EditResultAction
import com.mirash.familiar.tool.KEY_EDIT_RESULT_ACTION
import com.mirash.familiar.tool.KEY_ID
import com.mirash.familiar.tool.KEY_POSITION
import com.mirash.familiar.view.StyledTextInputLayout

/**
 * @author Mirash
 */
class CredentialsEditActivity : AppCompatActivity(), Observer<Credentials> {
    private lateinit var saveMenuItem: MenuItem

    private lateinit var binding: ActivityCredentialsEditBinding
    private lateinit var model: CredentialsEditViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCredentialsEditBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.credentialsToolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        model = ViewModelProvider(this)[CredentialsEditViewModel::class.java]
        intent.extras?.let {
            if (it.containsKey(KEY_ID)) {
                val id = it.getLong(KEY_ID, 0)
                model.setCredentialsId(id)
                model.credentialsLiveData?.observe(this, this)
            }
            val position = intent.getIntExtra(KEY_POSITION, 0)
            model.credentialsPosition = position
        }
        initSaveButtonStateObserver()
    }

    override fun onDestroy() {
        super.onDestroy()
        model.removeCredentialsObserver(this)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(
            if (model.credentialsId == null) R.menu.credentials_save_menu else R.menu.credentials_edit_menu, menu
        )
        saveMenuItem = menu.findItem(R.id.credentials_edit_save)
        model.saveButtonEnableStateLiveData.observe(this) { enabled: Boolean ->
            saveMenuItem.setEnabled(enabled)
            saveMenuItem.icon?.alpha = if (enabled) 255 else 78
        }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }

            R.id.credentials_edit_save -> {
                saveCredentials()
            }

            R.id.credentials_edit_delete -> {
                deleteCredentials()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onChanged(value: Credentials) {
        binding.itemTitleInput.setTextWithNoAnimation(value.title)
        binding.itemLinkInput.setTextWithNoAnimation(value.link)
        binding.itemLoginInput.setTextWithNoAnimation(value.login)
        binding.itemEmailInput.setTextWithNoAnimation(value.email)
        binding.itemPhoneInput.setTextWithNoAnimation(value.phone)
        binding.itemPasswordInput.setTextWithNoAnimation(value.password)
        binding.itemPinInput.setTextWithNoAnimation(value.pin)
        binding.itemDetailsInput.setTextWithNoAnimation(value.details)
    }

    private fun saveCredentials() {
        val credentialsModel = CredentialsModel()
        credentialsModel.title = binding.itemTitleInput.text
        credentialsModel.link = binding.itemLinkInput.text
        credentialsModel.login = binding.itemLoginInput.text
        credentialsModel.email = binding.itemEmailInput.text
        credentialsModel.phone = binding.itemPhoneInput.text
        credentialsModel.password = binding.itemPasswordInput.text
        credentialsModel.pin = binding.itemPinInput.text
        credentialsModel.details = binding.itemDetailsInput.text
        credentialsModel.position = model.credentialsPosition
        model.saveCredentials(credentialsModel)
        val data = Intent()
        data.putExtra(
            KEY_EDIT_RESULT_ACTION,
            if (model.credentialsId == null) EditResultAction.CREATE else EditResultAction.UPDATE
        )
        data.putExtra(KEY_POSITION, model.credentialsPosition)
        setResult(RESULT_OK, data)
        finish()
    }

    private fun deleteCredentials() {
        model.removeCredentialsObserver(this)
        model.deleteCredentials()
        val data = Intent()
        data.putExtra(KEY_POSITION, model.credentialsPosition)
        data.putExtra(KEY_EDIT_RESULT_ACTION, EditResultAction.DELETE)
        setResult(RESULT_OK, data)
        finish()
    }

    private fun initSaveButtonStateObserver() {
        initTextFillObserver(binding.itemTitleInput, CredentialsEditViewModel.INDEX_TITLE)
        initTextFillObserver(binding.itemLinkInput, CredentialsEditViewModel.INDEX_LINK)
        initTextFillObserver(binding.itemLoginInput, CredentialsEditViewModel.INDEX_LOGIN)
        initTextFillObserver(binding.itemEmailInput, CredentialsEditViewModel.INDEX_EMAIL)
        initTextFillObserver(binding.itemPhoneInput, CredentialsEditViewModel.INDEX_PHONE)
        initTextFillObserver(binding.itemPasswordInput, CredentialsEditViewModel.INDEX_PASSWORD)
        initTextFillObserver(binding.itemPinInput, CredentialsEditViewModel.INDEX_PIN)
    }

    private fun initTextFillObserver(inputLayout: StyledTextInputLayout, index: Int) {
        inputLayout.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun afterTextChanged(editable: Editable) {
                model.setFillState(index, editable.isNotEmpty())
            }
        })
    }
}
